### ISSUE TITLE
Extract to function (free function or member function)

### ISSUE THREAD

**Issue by JamesRamm** [microsoft/vscode-cpptools#1162]
Thanks for the great work on this; I am successfully using vscode-cpptools to work on a >2m line Qt project with no hitches so far :+1: :smile: 

Refactoring is a huge task on such a project - support for extracting a method or variable would be great!

I.e select a section of code -> right click -> extract method (keybinding would be great aswell)

**Comment by lambage** [comment]
*bump*

**Comment by hoffmael** [comment]
Advanced refactoring is highly desired for my workflow. This would bring C++ up to speed with other VS code languages that support the refactoring features listed here: https://code.visualstudio.com/docs/editor/refactoring

**Comment by Doppp** [comment]
Bump again.

**Comment by Audhumbla5** [comment]
bump - being able to refactor, extract methods and create definition from declaration is incredibly useful for efficiency.

**Comment by rockboynton** [comment]
bump

**Comment by fv2018** [comment]
bump

**Comment by GustavoOS** [comment]
Now that Visual Studio has Refactoring abilities for other languages, any chances to bring it to C++?

**Comment by ferdinandhelmer** [comment]
extract method, create declaration from implementation and change signature would be a *giant* plus. Great work on the C++ extension in general though, I am really liking it! Keep up the good spirit!

**Comment by nadya-p** [comment]
Moving a function from .h to .cpp (or in the opposite direction) is another example of useful C++ specific refactoring. Some nice C++ refactorings implemented in QtCreator are given [here](https://doc.qt.io/qtcreator/creator-editor-refactoring.html).

**Comment by n9wxu** [comment]
bump

**Comment by Lennon925** [comment]
@sean-mcmanus Any plan about this recently?

**Comment by sean-mcmanus** [comment]
@Lennon925 It's not on our plans yet -- it's our 10th most upvoted feature and we're currently working on some of the 9th and below most upvoted features for our 1.0 release (after 0.27.0).

**Comment by Lennon925** [comment]
> @Lennon925 It's not on our plans yet -- it's our 10th most upvoted feature and we're currently working on some of the 9th and below most upvoted features for our 1.0 release (after 0.27.0).

Thanks and Expect it very much~

**Comment by TherionIM** [comment]
bump

**Comment by stefanstrat** [comment]
bump

**Comment by FrightRisk** [comment]
Same here. Would live to highlight a section of code and have it extracted and turned it into a function.

**Comment by AnthonyParkssonos** [comment]
I had to take drastic measures to help with this and actually bought a CLion license. It is slow as dirt but the refactoring support is phenomenal. Would love to have this supported in VSCode which is my preferred IDE by far.

**Comment by rockboynton** [comment]
I have some time on my hands before starting on my internship and would like to contribute if I can. How would I go about trying to implement this? I haven't really contributed to open source projects other than my own but this could be a good start? 

My initial thought is to view the source for the languages where this feature is working (Java, Javascript, and others I think). 

Can anyone point me in the right direction?

Thanks!

**Comment by noamgot** [comment]
looking forward to such a feature - I use it a lot on PyCharm, and would love to see it on vscode as well

**Comment by sean-mcmanus** [comment]
@rockboynton Our open source TypeScript code doesn't have access to the database and lexing code in the closed source cpptools process that would be necessary to implement this correctly. If you wanted to use another lexer/parser, then implementing a separate extension that provides the extract method variable feature would be another possibility. The interface with VS Code is the code actions APIs.

**Comment by rockboynton** [comment]
@sean-mcmanus Thank you for the information. I will look into it.

**Comment by Vanuan** [comment]
@sean-mcmanus So you're saying `cpptools-srv` is closed source? Is it possible to run cpptools extension with `clangd` as LSP server? Or should we look into https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd ?

**Comment by sean-mcmanus** [comment]
@Vanuan Yes, cpptools-srv (and cpptools) is closed source and covered by a more restrictive [EULA](https://github.com/microsoft/vscode-cpptools/blob/master/Extension/LICENSE.txt), not the [MIT license](https://github.com/microsoft/vscode-cpptools/blob/master/LICENSE.md). Yeah, you can try https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd . We have not investigated invoking clangd from our extension.

**Comment by zbz-cool** [comment]
bump

**Comment by JensWehner** [comment]
bump

**Comment by guillermo-gerard** [comment]
Another bump :)

**Comment by BenthamWu** [comment]
one more bump:)

**Comment by malife** [comment]
Add on bump to the list :-)

**Comment by GustavoOS** [comment]
Come on, 3 years... it's getting ridiculous

**Comment by Tudyx** [comment]
Bump

**Comment by SvenSommer** [comment]
_Bump_

**Comment by OsamaRyees** [comment]
it is already 2021, can we Bump this

**Comment by lambage** [comment]
Features I'd like to see:

- Select a block of code, right click context menu with option to turn into function, then either a popup appears asking for a name or it does one of those cool in place box edits to give it a name.  All variables in the scope of the selection should be added as parameters to the function.  If only one non-class member variable is modified, the function should return that value, otherwise any modified non-class variables should be passed by reference.  If there are no class member variables it should make a free function in an anonymous namespace (similarly if the selected code scope is not within a class member function).  If the selected scope is within a class member and uses class member variables then the definition should be added to the cpp unit file and the declaration added to the header file (in a private section by default would suffice).
- Add a declaration to a header file, right click context menu with option to create definition in the cpp unit file, ideally in the same order as the header file.
- Add a definition to the cpp unit file, right click context menu with option to create declaration in the header file
- Context menu to sort definitions to match the order of declarations in the header file
- Context menu to cleanup headers and remove any unused headers (strecth goal here)
- Context menu to change order of parameters to a function

That's all I can think of for now.  Happy coding everyone!

**Comment by stefanofiorentino** [comment]
@jureid did this feature make the cut for the current milestone?

**Comment by sean-mcmanus** [comment]
@stefanofiorentino No, it's for 1.7.0 currently.

**Comment by atar-axis** [comment]
bump :) where can i upvote this feature?

**Comment by bobbrow** [comment]
Yes, clicking the 👍 is how you upvote.

**Comment by lkankowski** [comment]
Even simple extraction tool will be very helpfull. 👍

Btw: Great work!

**Comment by sampathrg** [comment]
This would be very useful. Refactoring source code is a pain without this feature. _Bump! Bump! Bump!_

**Comment by JoherYu** [comment]
Bump!

**Comment by stefanofiorentino** [comment]
I now export cmake project to ecplipse to have refactoring features... so sad.. (imho clangd is not powerful enough, while CLion is).

**Comment by claremacrae** [comment]
I'm unsubscribing from this, as such a basic refactoring still not being available after 4 years means I'm still using other tools now... and don't need notifications for all the understandable pings!

If this is ever implemented, I would appreciate a ping.

**Comment by GustavoOS** [comment]
This feature should be priority #1 for taking so long, with so many upvotes. Shame on who deprioritizes it.

**Comment by bilderbuchi** [comment]
> If this is ever implemented, I would appreciate a ping.

@claremacrae There is a "customize" setting by the Subscribe button, where you can set a notification only when this issue is closed ;-)

**Comment by atar-axis** [comment]
Bump :P tell me how I can help in making this one available, I know it's FOSS and therefore thank you for everything so far, but we would really appreciate it.

**Comment by viniciusalmada** [comment]
Bum again

**Comment by jacobjiggler** [comment]
Bumpity Bump

**Comment by rthickling** [comment]
Bump

**Comment by hillelme** [comment]
please make this happen :)

**Comment by naszim88** [comment]
Bump

**Comment by niksatgit** [comment]
Bump Bump Bump.. n upvotes and yet this is not winning. The voting is rigged! ;D 
Ok jokes apart, would love to see this feature.

**Comment by escherstair** [comment]
Bump

**Comment by soerenPeters** [comment]
Would be an extreme improvement for vscode. Heavy upvote!!!

**Comment by marcpawl** [comment]
Bump

**Comment by Johnlon** [comment]
+1 unbelievable 
VSCode is nothing but a glorified text editor without refactoring

**Comment by akochkarev** [comment]
+1
Refactoring tools for C++ is highly appreciated!

**Comment by JoshAGordon** [comment]
I thought I'd just missed something when I couldn't find refactoring. Can't believe this just isn't a feature in 2022!

**Comment by Johnlon** [comment]
Yeah it's astonishing. Vscode is a glorified text editor without it.

**Comment by davidzanger** [comment]
+1
It would be really helpful.

**Comment by Johnlon** [comment]
Wsl and makefiles seem to work better in vscode than visual studio which
does have basic refactoring.  but I need refactoring and makefiles and wsl
and I found clion a better outcome for what I'm doing.

**Comment by KatPurpy** [comment]
It's been almost 5 years since the issue was opened. Let's celebrate!

**Comment by ZA0068** [comment]
bump

**Comment by ZA0068** [comment]
Its been 5 years since the absence of this effective refactor tools.

Javascript and Python works well, but not C++?

this has to be done

**Comment by bobbrow** [comment]
It's on our roadmap.  We'll be looking at it after #664 is complete because this feature has a dependency on that one.

**Comment by kindtek** [comment]
In the past year I became a newfound believer in vs code. Ever since I installed it I have been trying to "debug" why my system won't do refactorings on c++. Thought it had to do with my WSL setup. Good to know there's nothing wrong with my system I guess. It would have been nice if it was more obvious on the refactoring help page or something. This would have saved me hours upon hours.

**Comment by Johnlon** [comment]
Yeah it's a joke. Makes vscode a glorified text editor rather than an ide.

**Comment by Stoff81** [comment]
Bump

**Comment by lvyaqiao** [comment]
I think there is something wrong with my c++ extension ,but at last i found it is coming soon……Maybe ”C++ does not support extract method" should be marked.
With this feature, vscode will be more awesome!

**Comment by abhijithnraj** [comment]
This will be a really good feature to have. Will definitely improve productivity.

**Comment by stefanofiorentino** [comment]
bump :) this is currently the only reason blocking the VS Code adoption in my projects. Refactoring is very important to us, and should be to you too.

**Comment by escherstair** [comment]
> It's on our roadmap. We'll be looking at it after #664 is complete because this feature has a dependency on that one.

Good news that at least #664 is available for preview.

**Comment by patrick-dojofive** [comment]
Eagerly anticipating this feature!

**Comment by winocm** [comment]
I would appreciate this feature, very, very much.

**Comment by vzagorovskiy** [comment]
bump

**Comment by crackwitz** [comment]
https://langserver.org/ lists at least three servers for "C/C++" code, LLVM being the most well known of them. I continue to hope that VSCode will eventually learn to deal with C and C++ code properly.

**Comment by stefanofiorentino** [comment]
> https://langserver.org/ lists at least three servers for "C/C++" code, LLVM being the most well known of them. I continue to hope that VSCode will eventually learn to deal with C and C++ code properly.

The clangd extension works quite well to me. The integration is most probably not feature-complete, but it's a good starting point (unsubscribing this, sorry).

**Comment by lluissm** [comment]
Please 🙏🏻

**Comment by starball5** [comment]
Related Stack Overflow question: [VS Code "Extract Method" and "Extract Variable" not in context menu for C++](https://stackoverflow.com/q/63446179/11107541)

**Comment by patrick-dojofive** [comment]
I would love to hear any updates about any in-progress work.

**Comment by sean-mcmanus** [comment]
@patrick-dojofive It's not in-progress -- it's assigned to project 1.16 and 1.15.0 is in progress.

**Comment by sean-mcmanus** [comment]
I've moved "Extract Variable" to a separate issue: https://github.com/microsoft/vscode-cpptools/issues/10930 .

**Comment by eraoul** [comment]
Wow -- this is 5+ years old and VSCode still doesn't support refactoring for C++? I used this feature like 2 decades ago in other IDEs. 

Is VSCode not intended for C++ use, and only really for C# or something? Super confused here.

**Comment by crackwitz** [comment]
From other open/closed issues, I would guess that VSC developers expect this to be done by extensions written by the community, through Language Server Protocol. There may also be an incentive to urge C/C++ developers towards buying Visual Studio instead (or using Eclipse with CDT...).

**Comment by stellarpower** [comment]
> The Visual Studio features were not written with linux/mac in mind
A meta-refactor - refactoring your refactorer! VS++

**Comment by sean-mcmanus** [comment]
FYI, we have a preliminary release of "Extract to function" available with 1.18.0 (https://github.com/microsoft/vscode-cpptools/releases/tag/v1.18.0), but it's only enabled when `C_Cpp.experimentFeatures` is set to `true`, and 'Extract to Free Function' is disabled, and it has a lot of known bugs we're working on fixing: https://github.com/microsoft/vscode-cpptools/issues?q=is%3Aopen+is%3Aissue+label%3A%22Feature%3A+Extract+to+function%22 -- also that list of known bugs is not complete (there others inherited from VS that haven't been added to GitHub yet).

0. Set `C_Cpp.experimentalFeatures` to `enabled`.
1. Select the code you want to extract to a function.
2. Select the "Extract to function" code action.
3. If the "Expand selection" code action appears, use that first.
4. If no code action appears, you can use Ctrl+Shift+R Ctrl+E to get an error message that may explain the reason for it not being available.
5. A rename prompt will appear, so use that, but we're planning to change it to prompt for the name before generating the code instead of doing a rename (https://github.com/microsoft/vscode-cpptools/issues/11531).
6. Afterwards the code will be formatted, but there are various issues that could cause formatting to not work, so we're looking into fixing those issues and/or formatting the code before it's generated (https://github.com/microsoft/vscode-cpptools/issues/10236).

**Comment by sean-mcmanus** [comment]
This is available in 1.18.4. C_Cpp.experimentalFeatures is not required. We're tracking some known issues: https://github.com/Microsoft/vscode-cpptools/issues?q=is%3Aopen+is%3Aissue+label%3A%22Feature%3A+Extract+to+function%22 (and there are other known issues that were previously reported on VS that haven't had issues added on GitHub yet).
